.crawler/src/crawler.rs

```diff
@@ -0,0 +1,274 @@
+                node.send_broadcast(ClientMessage::PeerRequest).unwrap();
+                tokio::time::sleep(Duration::from_secs(PEER_INTERVAL_SECS)).await;
+            }
+        });
+    }
+
+    fn log_known_network(&self) {
+        let node = self.clone();
+        tokio::spawn(async move {
+            loop {
+                info!(parent: node.node().span(), "Known addresses: {}", node.known_network.nodes().len());
+                info!(parent: node.node().span(), "Known connections: {}", node.known_network.connections().len());
+                tokio::time::sleep(Duration::from_secs(LOG_INTERVAL_SECS)).await;
+            }
+        });
+    }
+
+    /// Starts the usual periodic activities of a crawler node.
+    pub fn run_periodic_tasks(&self) {
+        self.send_pings();
+        self.update_peers();
+        self.log_known_network();
+    }
+}
+
+/// Inbound message processing logic for the crawler nodes.
+#[async_trait::async_trait]
+impl Reading for Crawler {
+    type Message = ClientMessage;
+
+    fn read_message<R: io::Read>(&self, source: SocketAddr, reader: &mut R) -> io::Result<Option<Self::Message>> {
+        // FIXME: use the maximum message size allowed by the protocol or (better) use streaming deserialization.
+        let mut buf = [0u8; 8 * 1024];
+
+        reader.read_exact(&mut buf[..MESSAGE_LENGTH_PREFIX_SIZE])?;
+        let len = u32::from_le_bytes(buf[..MESSAGE_LENGTH_PREFIX_SIZE].try_into().unwrap()) as usize;
+
+        if reader.read_exact(&mut buf[..len]).is_err() {
+            return Ok(None);
+        }
+
+        match ClientMessage::deserialize(&mut io::Cursor::new(&buf[..len])) {
+            Ok(msg) => {
+                info!(parent: self.node().span(), "received a {} from {}", msg.name(), source);
+                Ok(Some(msg))
+            }
+            Err(e) => {
+                error!("a message from {} failed to deserialize: {}", source, e);
+                Err(io::ErrorKind::InvalidData.into())
+            }
+        }
+    }
+
+    async fn process_message(&self, source: SocketAddr, message: Self::Message) -> io::Result<()> {
+        match message {
+            ClientMessage::BlockRequest(_start_block_height, _end_block_height) => {}
+            ClientMessage::BlockResponse(_block) => {}
+            ClientMessage::Disconnect(reason) => {
+                debug!("Peer {} disconnected for the following reason: {:?}", source, reason);
+            }
+            ClientMessage::PeerRequest => self.process_peer_request(source).await?,
+            ClientMessage::PeerResponse(peer_ips) => self.process_peer_response(source, peer_ips).await?,
+            ClientMessage::Ping(version, _fork_depth, _peer_type, _peer_state, _block_hash, block_header) => {
+                // Deserialise the block header.
+                let block_header = block_header.deserialize().await.unwrap();
```

Comment:
TODO: we should probably manually deserialize the header, as we only need the height, and we need to be able to quickly handle any number of such messages